File: src/main/java/algorithm/RobustSTL/robuststl.py
import sys
import numpy as np
from RobustSTL.RobustSTL import RobustSTL


def robuststl(y, T, reg1=1., reg2=0.5, K=2, H=5, dn1=1., dn2=1., ds1=50., ds2=1.):
    _, trend, seasonal, resid = RobustSTL(y, T, reg1=reg1, reg2=reg2, K=K, H=H, dn1=dn1, dn2=dn2, ds1=ds1, ds2=ds2)
    return trend, seasonal, resid

if __name__ == '__main__':
    T = 12
    if len(sys.argv) >= 2:
        T = int(sys.argv[1])  # First argument
    y = []
    f = open("/Users/chenzijie/Documents/GitHub/std-benchmark/src/main/java/algorithm/RobustSTL/input.txt","r")
    for line in f.readlines():
        if line.strip() != "":
            y.append(float(line.strip()))
    y = np.array(y)
    f.close()
    trend, seasonal, residual = robuststl(y, T)

    f = open("/Users/chenzijie/Documents/GitHub/std-benchmark/src/main/java/algorithm/RobustSTL/trend.txt","w")
    f.write("\n".join([str(item) for item in trend]))
    f.close()

    f = open("/Users/chenzijie/Documents/GitHub/std-benchmark/src/main/java/algorithm/RobustSTL/seasonal.txt","w")
    f.write("\n".join([str(item) for item in seasonal]))
    f.close()

    f = open("/Users/chenzijie/Documents/GitHub/std-benchmark/src/main/java/algorithm/RobustSTL/residual.txt","w")
    f.write("\n".join([str(item) for item in residual]))
    f.close()

File: src/main/java/algorithm/RobustSTL.java
package algorithm;

import java.io.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class RobustSTL {
    public final String FILE_DIR = "/Users/chenzijie/Documents/GitHub/std-benchmark/src/main/java/algorithm/RobustSTLUtils/";

    public Map<String, ArrayList<Double>> decompose(List<Double> ts, int period) throws Exception {
        ArrayList<Double> trend = new ArrayList<>();
        ArrayList<Double> seasonal = new ArrayList<>();
        ArrayList<Double> residual = new ArrayList<>();
        Map<String, ArrayList<Double>> result = new HashMap<>();

        // write
        StringBuilder string = new StringBuilder();
        for (double value : ts) {
            string.append(value);
            string.append("\n");
        }
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(FILE_DIR + "input.txt", false));
        bufferedWriter.write(string.toString());
        bufferedWriter.close();

        // python shell
        String[] cmd = {
                "/Users/chenzijie/anaconda3/envs/dq/bin/python",
                FILE_DIR + "robuststl.py",
                String.valueOf(period),
        };
        Process process = Runtime.getRuntime().exec(cmd);
        process.waitFor();

        if (process.exitValue() != 0){
            System.out.println("RobustSTL Complete UnSuccessfully.");
        }

        String line;

//        BufferedReader stdInput = new BufferedReader(new InputStreamReader(process.getInputStream()));
//        BufferedReader stdError = new BufferedReader(new InputStreamReader(process.getErrorStream()));
//        BufferedReader input = new BufferedReader(new InputStreamReader(process.getInputStream()));
//        while ((line = input.readLine()) != null) {
//            System.out.println(line);
//        }
//        input.close();

        // read
        BufferedReader bufferedReader = new BufferedReader(new FileReader(FILE_DIR + "trend.txt"));
        while ((line = bufferedReader.readLine()) != null) {
            trend.add(Double.parseDouble(line));
        }
        bufferedReader.close();

        bufferedReader = new BufferedReader(new FileReader(FILE_DIR + "seasonal.txt"));
        while ((line = bufferedReader.readLine()) != null) {
            seasonal.add(Double.parseDouble(line));
        }
        bufferedReader.close();

        bufferedReader = new BufferedReader(new FileReader(FILE_DIR + "residual.txt"));
        while ((line = bufferedReader.readLine()) != null) {
            residual.add(Double.parseDouble(line));
        }
        bufferedReader.close();

        // record
        result.put("trend", trend);
        result.put("seasonal", seasonal);
        result.put("residual", residual);

        return result;
    }
}